File: Java/Intermediate/boj_1956.java
package Intermediate;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_1956 {
    private static final int INF = (int) 1e9;
    private static final BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));

    private static StringTokenizer st;
    private static int[][] grp;


    private static int solve() {
        for (int k = 1; k < grp.length; k++)
            for (int i = 1; i < grp.length; i++)
                for (int j = 1; j < grp.length; j++)
                    grp[i][j] = Math.min(grp[i][j], grp[i][k] + grp[k][j]);

        int ans = INF;

        for (int i = 1; i < grp.length; i++) {
            for (int j = 1; j < grp.length; j++) {
                if (i == j) continue;
                ans = Math.min(ans, grp[i][j] + grp[j][i]);
            }
        }

        return ans == INF ? -1 : ans;
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());

        grp = new int[v + 1][v + 1];
        for (int i = 1; i < grp.length; i++)
            Arrays.fill(grp[i], INF);

        while (e-- > 0) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            grp[a][b] = c;
        }

        System.out.println(solve());
    }
}

File: Python/Intermediate/boj_1956.py
import sys

INF = int(1e9)


def solve(grp: list[list[int]]) -> int:
    for k in range(1, len(grp)):
        for i in range(1, len(grp)):
            for j in range(1, len(grp)):
                grp[i][j] = min(grp[i][j], grp[i][k] + grp[k][j])

    ans = INF

    for i in range(1, len(grp)):
        for j in range(1, len(grp)):
            if i == j: continue
            ans = min(ans, grp[i][j] + grp[j][i])

    return -1 if ans == INF else ans


if __name__ == "__main__":
    input = sys.stdin.readline
    v, e = map(int, input().split())

    grp = [[INF] * (v + 1) for _ in range(v + 1)]
    for _ in range(e):
        a, b, c = map(int, input().split())
        grp[a][b] = c

    print(solve(grp))
